### PR TITLE
fix: numpy 2.0 compatibility

### PR DESCRIPTION
**Pull request by lgray**
This uses latest numpy in the build system, oldest possible numpy is no longer a correct choice.

Checked on laptop with numpy 1 and this package built against numpy 2.

### CODE REVIEW

**Comment by lgray** [comment]
@tpgillam whenever you get a chance to pull this in and make a new release, it would be appreciated!

**Comment by tpgillam** [comment]
Also, as far as I can tell, `oldest-supported-numpy` is still maintained and recommended.

https://pypi.org/project/oldest-supported-numpy/

What breaks for you with it?

**Comment by lgray** [comment]
https://github.com/scipy/oldest-supported-numpy

Please have a look at the deprecation notice on their GitHub. It might be on pypi but it is no longer the right way to do things, and does indeed cause problems. There's no forwards compatibility.

**Comment by lgray** [comment]
It looks like your oldest supported numpy is going to be 1.26, given that you would like to support python 3.12.

**Comment by lgray** [comment]
@tpgillam that should be most of the spring cleaning :-)

**Comment by lgray** [comment]
The only thing I need to add is to test with numpy 2 and numpy 1

**Comment by tpgillam** [comment]
Just drive-by commenting, don't have time to look in detail at your work right now, sorry!

> https://github.com/scipy/oldest-supported-numpy

Thanks for the reference! I'll read this in more detail later.

> It looks like your oldest supported numpy is going to be 1.26, given that you would like to support python 3.12.

I feel like it should be possible to support older versions when using a suitable python interpreter. I'm inclined to be as protective as possible towards users that might be on older systems.

**Comment by lgray** [comment]
and there we go.

**Comment by lgray** [comment]
The answer that users will get from the python core developers for anything less than python 3.9 is "that version of python is no longer supported". You're pretty well situated to rattle things around and tell people to update their systems.

**Comment by lgray** [comment]
Ah lovely - I can fix up the problems so long as you let the CI run. Thanks!

**Comment by Rupt** [comment]
Thanks for your diligent contributions @lgray.

I can't comment on the implementation, but your statements about NumPy compatibility surprise me, because they contradict NumPy's support specification https://scientific-python.org/specs/spec-0000/ (forwarded from https://numpy.org/neps/nep-0029-deprecation_policy.html).

I can sadly confirm that you are correct: NumPy versions <1.26 does not support Python 3.12:

```shell
PYTHON_BUILD=cpython-3.12.4+20240726-x86_64-unknown-linux-gnu-install_only_stripped.tar.gz
wget https://github.com/indygreg/python-build-standalone/releases/download/20240726/${PYTHON_BUILD}
tar xvf ${PYTHON_BUILD}
python/bin/python --version  # Python 3.12.4
python/bin/pip install numpy==2.0  # ok
python/bin/pip install numpy==1.26  # ok
python/bin/pip install numpy==1.25  # error
python/bin/pip install numpy==1.24  # error
```

**Comment by lgray** [comment]
Ah, you're right @Rupt with respect to the numpy support window. The issue comes when you want to support 2 and 3.12 simultaneously. I don't have a machine where I can test numpy 2 vs. much older numpy (arm macos machine right now), but if you *really* need it we can support back to 1.22 (and I *think* you can do that compiling with 2). 

However, there's not too much of a good reason to do that, IMO, considering current environment management practices.

File: src/mt2_Lallyver2.h
Comment (tpgillam) [inline src/mt2_Lallyver2.h:536]: Can you revert changes in this file please? It's a reference implementation from a third party, so we don't want to change it (at least not in-place).
Recommended implementation for now is `mt2_bisect.h`

Thanks!

File: .github/workflows/ci.yml
Comment (tpgillam) [inline .github/workflows/ci.yml:45]:
```suggestion
                    - os: windows-latest
```
I think, otherwise duplicated below.

File: .github/workflows/ci.yml
Comment (tpgillam) [inline .github/workflows/ci.yml:41]: Do we need to do some thinking to ensure that wheels are built for macos x86?

Looks like we should be able to cross compile, but might mean setting `universal2` as the arch for the macos job. Just skimmed:
- https://cibuildwheel.pypa.io/en/stable/options/#archs
- https://cibuildwheel.pypa.io/en/stable/faq/#apple-silicon

File: setup.py
Comment (tpgillam) [inline setup.py:51]: Looking at these docs, they suggest `1.23.5` should be supportable at least:

https://numpy.org/doc/stable/dev/depending_on_numpy.html#numpy-2-0-specific-advice

Since it's just a lower bound, it should be fine that on e.g. python 3.12 this bound will be superseded.

Comment (lgray) [reply setup.py:51]: Sure - lemme do that and integrate that sort of info into the tests.

Comment (tpgillam) [reply setup.py:51]: Excellent, thanks!

Somewhat related; I think we need to force the build to happen with numpy2. Example uses the `[build-system]` section of pyproject.toml. Not sure what the old-style incantation is.

Migrating to `pyproject.toml` is on the bucket list, but hopefully not required to do right now.

Comment (lgray) [reply setup.py]: Nah - migrating to `pyproject.toml` is not necessary (but would be easy and make some of your deployment stuff much more clean). 

A clearly useful migration would be to use pybind11 instead of doing it all yourself!

Comment (tpgillam) [reply setup.py]: > A clearly useful migration would be to use pybind11 instead of doing it all yourself!

I was & am keen to keep the dependency set as light as possible for this package. It's very simple, and no new features are planned. (Was a conscious decision to avoid pybind or other wrapper-layers when writing originally!)

Comment (lgray) [reply setup.py]: It's header only C++, and manages the interface to numpy in a way that maintains all the backwards compatibility. So it really is minimal and reduces maintenance work. 

Numpy 2 compatibility would have come for free just switching to 2.12. Much less pain in the long run (at least with the package's maturity as it is now).

**Comment by tpgillam** [comment]
@lgray thanks for your continuing efforts to get the CI working! Wondering whether it would be easier for you to create a PR into your own fork such that you can run the CI yourself for testing (at least I _presume_ that would work)? I'm happy to press 'approve & run' when I see it, but maybe you'd prefer to have faster iteration times.

File: setup.py
Comment (tpgillam) [reply setup.py]: Have you used pybind11 to create ufuncs before? Thinking back I seem to recall that it wasn't supported. Actually not sure that it is now from a skim of the source code?

Anyway, even if it were possible I've actually been pleasantly surprised that no code changes have been required to get it building! Seems to be isolated to fiddly build system stuff.

Comment (lgray) [reply setup.py]: That was indeed lucky!

**Comment by lgray** [comment]
A bunch of the failures in the wheel CI are from the wheels already existing since you seem to be uploading even in PR! A little dangerous.

I'll make it check for a release tag.

**Comment by lgray** [comment]
Oh, nevermind, that's just the github artifacts.

**Comment by tpgillam** [comment]
> I'll make it check for a release tag.

We have this check:

https://github.com/tpgillam/mt2/blob/646818cf1aa95b9a32126abf6582e0438648121e/.github/workflows/ci.yml#L110

maybe it can be made better.

**Comment by lgray** [comment]
@tpgillam could you have a look in github and see if you can kill the uploaded artifacts for this PR?

**Comment by lgray** [comment]
I've changed the CI so it only uploads artifacts when there's a version tag. So I think this is almost there.

**Comment by lgray** [comment]
ok - nice - it works back to numpy 1.19.3 when built against numpy 2, according to the tests in python 3.8. Have adjusted all the pins.

**Comment by lgray** [comment]
Indeed, running the CI over on my own fork everything passes now, and we correctly test numpy back to oldest supported in 3.9.

https://github.com/lgray/mt2/actions/runs/10186047164/job/28176969069?pr=1

**Comment by tpgillam** [comment]
Thanks @lgray . github doesn't list any persisted artifacts for this workflow, so let's see what happens in this CI run.

Could you please bump the version in all appropriate places to 1.2.1 and add entry in HISTORY so that we can then tag once merged?

File: setup.py
Comment (tpgillam) [inline setup.py:52]: Instead of deleting entirely, let's update to indicate 3.9-3.12 support.

File: .github/workflows/ci.yml
Comment (tpgillam) [inline .github/workflows/ci.yml:28]: I think isn't required, since these lines are also included in the `tox.ini`

**Comment by lgray** [comment]
looks like it's all good to go aside from your github options expecting a job simply called "Build"!

**Comment by lgray** [comment]
OK - @tpgillam added two things and I think this is done.

1) `Build / pass` job label represents all jobs passing, you'll need to add that to your repository config
2) as commits are added to a PR if there are previously running CI jobs on the same PR it will cancel them, better for GHA resources

**Review by tpgillam — APPROVED** [review]
Thanks @lgray , this looks great!

The required checks on github are baffling me. I wish they were configurable more sensibly, but hey-ho. I've tried but so far failed to get it to match `Build / pass` or `Build / pass (pull request)`.

edit: apparently the magic was just to require `pass` on the github side.

**Comment by lgray** [comment]
Cool, thanks a lot for dealing with all the churn. :-)

Whenever you're ready, please merge and make a release!

**Pull request merged**
